### ISSUE TITLE
Make MF6 callable, i.e. live modifications of MF6 values with Python

### ISSUE THREAD

**Issue by pya** [MODFLOW-USGS/modflow6#180]
# Motivation

The objective is to make MODFLOW 6 (MF6) callable. This means the model user is able to automatically interact with a running model, using a user-written Python program.

# Proposed Solution

The package [pymf6](https://github.com/hydrocomputing/pymf6) is a proof-of-concept implementation of a Python solution. The [documentation](https://pymf6.readthedocs.io/en/latest/index.html) explains how it works. The [principle](https://pymf6.readthedocs.io/en/latest/principle.html) shows how the user can change MF6 values at run time. The [current way of building](https://pymf6.readthedocs.io/en/latest/developer.html) depends on this [fork](https://github.com/hydrocomputing/modflow6) of MF6. 

# Needed Changes to the MF6 source

In order to make `pymf6` work with the source of this MF6 repo the following changes are necessary.

## 1. Move the code of the main program into a subroutine

The main program `mf6.f90` will call the subroutine `mf6_sub_outer`:

```fortran
program mf6

  call mf6_sub_outer()

end program mf6
```

## 2. Add a callback

The subroutine `mf6_sub_outer` in the file `mf6sub.f90`:

```fortran
subroutine mf6_sub_outer()

  use :: KindModule, only: cback

  call mf6_sub(cback)

end subroutine mf6_sub_outer
```

It calls the subroutine `mf6_sub` with an callback subroutine that comes from the module `KindModule`.

The subroutine `mf6_sub` is also located in `mf6sub.f90`:

```
subroutine mf6_sub(cback)

... 

tsloop: do
    !
    ! -- TIME UPDATE (TU)
    call tdis_tu()
    call cback()


....

enddo
```

It contains the original code from the main program. The only difference to the original code is the call to `cback()` at the beginning of of each time step loop. In `pymf6` this callback is a Python callable. In the standalone version `cback` is an external Fortran subroutine.

## Add an empty callback subroutine

The file `kind.f90` contains this code:

```
module KindModule

  implicit none

  public

...

  contains
....

  subroutine cback()
    ! empty

  end subroutine cback

end module KindModule
```

The subroutine `cback` is empty. This subroutine can be located in a different module as long this module is compiled at the very beginning of the build process. The new call to this empty subroutine should not add any run time, as the compiler will like inline this subroutine and therefore no additional instructions will be executed.

**Comment by pya** [comment]
I made a new release of `pymf6`, i.e. "callable MF6". Now you can inspect all variables the memory manger exposes interactively in a Jupyter Notebook. See [this Notebook](https://pymf6.readthedocs.io/en/latest/notebooks/Interactive.html) for an example.

Furthermore, I made the usage simpler. There is also a new [explaination](https://pymf6.readthedocs.io/en/latest/principle.html)  how it works.

Two [non-interactive examples](https://pymf6.readthedocs.io/en/latest/examples.html) show what can be done with it.

Let me know what you think about it. Is this going into a direction you imagined?

**Comment by pya** [comment]
I put some examples up on mybinder.org. You [try](https://mybinder.org/v2/gh/hydrocomputing/pymf6-examples/master?urlpath=lab/tree/index.ipynb) how `pymf6` works without installing it.

**Comment by langevin-usgs** [comment]
The MODFLOW 6 source code has been modified to adopt a BMI standard.  Our first implementation will be released in 6.1.1.

**Comment by pya** [comment]
Great news. So 6.1.1. will be the next release.
